Test instantiation exception with multicatch

**Describe the bug**

I have code that uses a multi-catch , which works nicely with `mvn compile quarkus:dev`, but fails when stating Quarkus as part of a test with `@Quarkustest`

The server shows

```
org.junit.jupiter.api.extension.TestInstantiationException: TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [com.redhat.cloud.custompolicies.engine.PolicyEvalTest]
Caused by: java.lang.ExceptionInInitializerError
Caused by: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.RuntimeException: Failed to initialize Arc
Caused by: java.lang.VerifyError: 
Bad type on operand stack
Exception Details:
  Location:
    com/redhat/cloud/custompolicies/engine/RestApi.verifyPolicy(Lcom/redhat/cloud/custompolicies/engine/model/Policy;)Ljavax/ws/rs/core/Response; @22: invokevirtual
  Reason:
    Type 'java/lang/Object' (current frame, stack[2]) is not assignable to 'java/lang/RuntimeException'
  Current Frame:
    bci: @22
    flags: { }
    locals: { 'com/redhat/cloud/custompolicies/engine/RestApi', 'com/redhat/cloud/custompolicies/engine/model/Policy', top, 'java/lang/Object' }
    stack: { uninitialized 17, uninitialized 17, 'java/lang/Object' }
  Bytecode:
    0x0000000: bb00 0a59 2bb7 0014 57b8 000c 4da7 0021
    0x0000010: 4ebb 0016 592d b600 17b7 0018 3a04 1101
    0x0000020: 9c2d b600 17b8 0019 1904 b600 0e4d 2c12
    0x0000030: 1b12 1cb6 001d 572c b600 0fb0          
  Exception Handler Table:
    bci [0, 13] => handler: 16
    bci [0, 13] => handler: 16
  Stackmap Table:
    same_locals_1_stack_item_frame(@16,Object[#56])
    append_frame(@46,Object[#118])
```

Source is like (full code is at https://github.com/pilhuhn/custom-policies-engine/blob/test-sys-profile/src/main/java/com/redhat/cloud/custompolicies/engine/RestApi.java#L93 )

```
  public Response verifyPolicy(Policy policy) {

    Response.ResponseBuilder builder;
    try {
      new PolicyParser(policy);
      builder = Response.noContent();
//    } catch (Exception ie) {
    } catch (IllegalArgumentException|IllegalStateException ie) {
      Msg msg = new Msg(ie.getMessage());
```

When I replace the line with multi-catch with the simple catch in the commented line above,
all works normally.

This is with 1.0.0.Final and .CR2

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
Darwin snert 18.7.0 Darwin Kernel Version 18.7.0: Sat Oct 12 00:02:19 PDT 2019; root:xnu-4903.278.12~1/RELEASE_X86_64 x86_64

 - Output of `java -version`: 

java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)

 - GraalVM version (if different from Java): 

not used.

Similar, but no multicatch. Again: code runs nicely in quarkus:dev mode

```
Caused by: java.lang.RuntimeException: Failed to register class com.redhat.cloud.custompolicies.app.EngineResponseExceptionMapper
        at org.jboss.resteasy.microprofile.client.RestClientBuilderImpl.newInstanceOf(RestClientBuilderImpl.java:456)
        at org.jboss.resteasy.microprofile.client.RestClientBuilderImpl.register(RestClientBuilderImpl.java:469)
        at org.jboss.resteasy.microprofile.client.RestClientBuilderImpl.build(RestClientBuilderImpl.java:206)
        at io.quarkus.restclient.runtime.RestClientBase.create(RestClientBase.java:57)
        at com.redhat.cloud.custompolicies.app.VerifyEngine_6a906821d45dd8b81c807f4d2493863b2bc24c83_Synthetic_Bean.create(VerifyEngine_6a906821d45dd8b81c807f4d2493863b2bc24c83_Synthetic_Bean.zig:34)
        at com.redhat.cloud.custompolicies.app.VerifyEngine_6a906821d45dd8b81c807f4d2493863b2bc24c83_Synthetic_Bean.get(VerifyEngine_6a906821d45dd8b81c807f4d2493863b2bc24c83_Synthetic_Bean.zig:207)
        at com.redhat.cloud.custompolicies.app.VerifyEngine_6a906821d45dd8b81c807f4d2493863b2bc24c83_Synthetic_Bean.get(VerifyEngine_6a906821d45dd8b81c807f4d2493863b2bc24c83_Synthetic_Bean.zig:62)
        at io.quarkus.arc.impl.CurrentInjectionPointProvider.get(CurrentInjectionPointProvider.java:53)
        at com.redhat.cloud.custompolicies.app.rest.PolicyCrudService_Bean.create(PolicyCrudService_Bean.zig:367)
        ... 30 more
Caused by: java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    com/redhat/cloud/custompolicies/app/EngineResponseExceptionMapper.toThrowable(Ljavax/ws/rs/core/Response;)Ljava/lang/RuntimeException; @78: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'java/lang/RuntimeException' (from method signature)
  Current Frame:
    bci: @78
    flags: { }
    locals: { 'com/redhat/cloud/custompolicies/app/EngineResponseExceptionMapper', 'javax/ws/rs/core/Response', integer, 'com/redhat/cloud/custompolicies/app/model/Msg', 'java/lang/Object' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0x0000000: 2bb6 0002 3d2a 2bb7 0003 4e1c ab00 0000
    0x0000010: 0000 0036 0000 0001 0000 019c 0000 0014
    0x0000020: bb00 0459 bb00 0559 b700 0612 07b6 0008
    0x0000030: 2db4 0009 b600 08b6 000a b700 0b3a 04a7
    0x0000040: 000d bb00 0c59 1cb7 000d 3a04 1904 b0  
  Stackmap Table:
    append_frame(@32,Integer,Object[#45])
    same_frame(@66)
    append_frame(@76,Object[#25])

        at java.lang.Class.getDeclaredConstructors0(Native Method)
        at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
        at java.lang.Class.getConstructor0(Class.java:3075)
        at java.lang.Class.newInstance(Class.java:412)
        at org.jboss.resteasy.microprofile.client.RestClientBuilderImpl.newInstanceOf(RestClientBuilderImpl.java:454)
        ... 38 more

```

--- a/core/deployment/src/main/java/io/quarkus/deployment/QuarkusClassWriter.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/QuarkusClassWriter.java
@@ -20,21 +20,8 @@ public class QuarkusClassWriter extends ClassWriter {
     }
 
     @Override
-    protected String getCommonSuperClass(String type1, String type2) {
-        ClassLoader cl = getClassLoader();
-        Class<?> c1 = null, c2 = null;
-        try {
-            c1 = cl.loadClass(type1.replace('/', '.'));
-        } catch (ClassNotFoundException e) {
-        }
-        try {
-            c2 = cl.loadClass(type2.replace('/', '.'));
-        } catch (ClassNotFoundException e) {
-        }
-        if (c1 != null && c2 != null) {
-            return super.getCommonSuperClass(type1, type2);
-        }
-        return Object.class.getName().replace('.', '/');
+    protected ClassLoader getClassLoader() {
+        // the TCCL is safe for transformations when this ClassWriter runs
+        return Thread.currentThread().getContextClassLoader();
     }
-
 }

--- a/core/deployment/src/main/java/io/quarkus/runner/RuntimeRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/RuntimeRunner.java
@@ -29,6 +29,7 @@ import io.quarkus.deployment.QuarkusAugmentor;
 import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
 import io.quarkus.deployment.builditem.ApplicationClassNameBuildItem;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
+import io.quarkus.deployment.builditem.DeploymentClassLoaderBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.LiveReloadBuildItem;
@@ -118,7 +119,14 @@ public class RuntimeRunner implements Runnable, Closeable {
                     functions.computeIfAbsent(i.getClassToTransform(), (f) -> new ArrayList<>()).add(i.getVisitorFunction());
                 }
 
+                DeploymentClassLoaderBuildItem deploymentClassLoaderBuildItem = result
+                        .consume(DeploymentClassLoaderBuildItem.class);
+                ClassLoader previous = Thread.currentThread().getContextClassLoader();
+
+                // make sure we use the DeploymentClassLoader for executing transformers since this is the only safe CL for transformations at this point
+                Thread.currentThread().setContextClassLoader(deploymentClassLoaderBuildItem.getClassLoader());
                 transformerTarget.setTransformers(functions);
+                Thread.currentThread().setContextClassLoader(previous);
             }
 
             if (loader instanceof RuntimeClassLoader) {

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -190,7 +190,8 @@ public class QuarkusTestExtension
 
                                         @Override
                                         protected ClassLoader getClassLoader() {
-                                            return temp;
+                                            // this has been previously set to a safe for transformations CL
+                                            return main;
                                         }
                                     };
                                     ClassLoader old = Thread.currentThread().getContextClassLoader();

